File: test/fixtures/samples/auth-large.go
package auth

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"
)

func Authenticate(w http.ResponseWriter, r *http.Request) {
	token := r.Header.Get("Authorization")
	if token == "" {
		http.Error(w, "No token provided", http.StatusUnauthorized)
		return
	}
	decoded, err := ValidateToken(token)
	if err != nil {
		http.Error(w, "Invalid token", http.StatusUnauthorized)
		return
	}
	if decoded.Exp < time.Now().Unix() {
		http.Error(w, "Token expired", http.StatusUnauthorized)
		return
	}
	if len(decoded.Permissions) == 0 {
		http.Error(w, "Insufficient permissions", http.StatusForbidden)
		return
	}
	userAgent := r.UserAgent()
	ipAddress := r.RemoteAddr
	fmt.Printf("Auth success: user=%s ip=%s ua=%s\n", decoded.Sub, ipAddress, userAgent)
	w.WriteHeader(http.StatusOK)
}

func ValidateToken(token string) (*Claims, error) {
	if token == "" {
		return nil, fmt.Errorf("empty token")
	}
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid token format")
	}
	headerBytes, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return nil, err
	}
	var header map[string]interface{}
	if err := json.Unmarshal(headerBytes, &header); err != nil {
		return nil, err
	}
	if header["alg"] != "HS256" {
		return nil, fmt.Errorf("unsupported algorithm")
	}
	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, err
	}
	var claims Claims
	if err := json.Unmarshal(payloadBytes, &claims); err != nil {
		return nil, err
	}
	return &claims, nil
}

func RefreshSession(session *Session, timeout int64) (*SessionResult, error) {
	if session == nil || session.ID == "" {
		return nil, fmt.Errorf("invalid session")
	}
	now := time.Now().Unix()
	elapsed := now - session.LastAccess
	if elapsed > timeout {
		return &SessionResult{Expired: true}, nil
	}
	session.LastAccess = now
	session.RefreshCount++
	if session.RefreshCount > 100 {
		session.Flags = append(session.Flags, "high-refresh-rate")
	}
	remaining := timeout - elapsed
	return &SessionResult{
		Expired:      false,
		Session:      session,
		Remaining:    remaining,
		RefreshCount: session.RefreshCount,
	}, nil
}

File: test/fixtures/samples/auth-large.rs
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};

pub fn authenticate(headers: &HashMap<String, String>) -> Result<AuthResult, AuthError> {
    let token = headers.get("Authorization")
        .ok_or(AuthError::NoToken)?;

    let decoded = validate_token(token)?;

    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();

    if decoded.exp < now {
        return Err(AuthError::TokenExpired);
    }

    if decoded.permissions.is_empty() {
        return Err(AuthError::InsufficientPermissions);
    }

    let user_agent = headers.get("User-Agent")
        .map(|s| s.as_str())
        .unwrap_or("unknown");
    let ip_address = headers.get("X-Forwarded-For")
        .map(|s| s.as_str())
        .unwrap_or("unknown");

    println!("Auth success: user={} ip={} ua={}", decoded.sub, ip_address, user_agent);
    Ok(AuthResult { authenticated: true, user: decoded.sub })
}

pub fn validate_token(token: &str) -> Result<Claims, AuthError> {
    if token.is_empty() {
        return Err(AuthError::InvalidToken);
    }
    let parts: Vec<&str> = token.split('.').collect();
    if parts.len() != 3 {
        return Err(AuthError::InvalidToken);
    }
    let header_bytes = base64_decode(parts[0])
        .map_err(|_| AuthError::InvalidToken)?;
    let header: HashMap<String, String> = serde_json::from_slice(&header_bytes)
        .map_err(|_| AuthError::InvalidToken)?;
    if header.get("alg").map(|s| s.as_str()) != Some("HS256") {
        return Err(AuthError::UnsupportedAlgorithm);
    }
    let payload_bytes = base64_decode(parts[1])
        .map_err(|_| AuthError::InvalidToken)?;
    let claims: Claims = serde_json::from_slice(&payload_bytes)
        .map_err(|_| AuthError::InvalidToken)?;
    Ok(claims)
}

pub fn refresh_session(session: &mut Session, timeout: u64) -> SessionResult {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();
    let elapsed = now - session.last_access;
    if elapsed > timeout {
        return SessionResult { expired: true, session: None, remaining: 0, refresh_count: 0 };
    }
    session.last_access = now;
    session.refresh_count += 1;
    if session.refresh_count > 100 {
        session.flags.push("high-refresh-rate".to_string());
    }
    let remaining = timeout - elapsed;
    SessionResult {
        expired: false,
        session: Some(session.clone()),
        remaining,
        refresh_count: session.refresh_count,
    }
}
